aws/ami.go

```diff
@@ -0,0 +1,69 @@
+package aws
+
+import (
+	"time"
+
+	awsgo "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/gruntwork-io/aws-nuke/logging"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+// Returns a formatted string of AMI Image ids
+func getAllAMIs(session *session.Session, region string, excludeAfter time.Time) ([]*string, error) {
+	svc := ec2.New(session)
+
+	params := &ec2.DescribeImagesInput{
+		Owners: []*string{awsgo.String("self")},
+	}
+
+	output, err := svc.DescribeImages(params)
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	var imageIds []*string
+	for _, image := range output.Images {
+		layout := "2006-01-02T15:04:05.000Z"
+		createdTime, err := time.Parse(layout, *image.CreationDate)
+		if err != nil {
+			return nil, err
+		}
+
+		if excludeAfter.After(createdTime) {
+			imageIds = append(imageIds, image.ImageId)
+		}
+	}
+
+	return imageIds, nil
+}
+
+// Deletes all AMIs
+func nukeAllAMIs(session *session.Session, imageIds []*string) error {
+	svc := ec2.New(session)
+
+	if len(imageIds) == 0 {
+		logging.Logger.Infof("No AMIs to nuke in region %s", *session.Config.Region)
+		return nil
+	}
+
+	logging.Logger.Infof("Terminating all AMIs in region %s", *session.Config.Region)
+
+	for _, imageID := range imageIds {
+		params := &ec2.DeregisterImageInput{
+			ImageId: imageID,
+		}
+
+		_, err := svc.DeregisterImage(params)
+		if err != nil {
```

Comment:
If you try to delete an AMI that's actively in use, do you get an error? If so, we should catch it, log it, and keep going.

Reply:
Will try this out

Reply:
From my tests an AMI can be deleted even when there exists an EC2 instance that was created with it. I believe once the VM is provisioned with that machine image there's no link between them and that ami is exists just for re-provisioning new EC2 instances

Reply:
Good to know!

Ship it 👍

Reply:
Sweet, I'll merge once I fix the test

Reply:
@tonerdo Is deleting an AMI actually a two-step procedure as [I had to do](https://github.com/josh-padnick/ec2-snapper/blob/master/delete_command.go#L258) in ec2-snapper? At that time, I first had to deregister the AMI (as you do here), and second, had to delete the AMI's corresponding EBS snapshot.